nupic/research/frameworks/quantization/qat_kwinners.py

```diff
@@ -0,0 +1,168 @@
+            n=n,
+            percent_on=percent_on,
+            k_inference_factor=k_inference_factor,
+            boost_strength_factor=boost_strength_factor,
+            duty_cycle_period=duty_cycle_period,
+        )
+        self.qconfig = qconfig
+        self.activation_post_process = qconfig.activation()
+
+    def forward(self, x):
+        if self.training:
+            x = F.KWinners.apply(x, self.duty_cycle, self.k, self.boost_strength)
+            self.update_duty_cycle(x)
+        else:
+            x = F.KWinners.apply(x, self.duty_cycle, self.k_inference,
+                                 self.boost_strength)
+
+        return self.activation_post_process(x)
+
+    @classmethod
+    def from_float(cls, mod, qconfig=None):
+        r"""Create a qat module from a float module or qparams_dict
+
+            Args: `mod` a float module, either produced by torch.quantization utilities
+            or directly from user
+        """
+        assert type(mod) == cls._FLOAT_MODULE, + cls.__name__ + \
+            ".from_float only works for " + cls._FLOAT_MODULE.__name__
+        if not qconfig:
+            assert hasattr(mod, "qconfig"), \
+                "Input float module must have qconfig defined"
+            assert mod.qconfig, "Input float module must have a valid qconfig"
+            qconfig = mod.qconfig
+
+        qat_kwinners = cls(
+            n=mod.n,
+            percent_on=mod.percent_on,
+            k_inference_factor=mod.k_inference_factor,
+            boost_strength_factor=mod.boost_strength_factor,
+            duty_cycle_period=mod.duty_cycle_period,
+            qconfig=qconfig
+        )
+        qat_kwinners.boost_strength = mod.boost_strength
+        qat_kwinners.duty_cycle = mod.duty_cycle
+        return qat_kwinners
+
+
+class QATKWinners2d(KWinners2d):
+    """
+    Similar to nupic.torch.modules.KWinners, with FakeQuantize modules
+    initialized to default.
+
+    Attributes:
+        activation_post_process: fake quant module for output activation
+    """
+    _FLOAT_MODULE = KWinners2d
+
+    def __init__(self, channels, percent_on, k_inference_factor,
+                 boost_strength_factor, duty_cycle_period, local, qconfig):
+
+        super().__init__(
+            channels=channels,
+            percent_on=percent_on,
+            k_inference_factor=k_inference_factor,
+            boost_strength_factor=boost_strength_factor,
+            duty_cycle_period=duty_cycle_period,
+            local=local
+        )
+
+        self.qconfig = qconfig
+        self.activation_post_process = qconfig.activation()
+
+    def forward(self, x):
+        if self.n == 0:
+            self.n = np.prod(x.shape[1:])
+            if not self.local:
+                self.k = int(round(self.n * self.percent_on))
+                self.k_inference = int(round(self.n * self.percent_on_inference))
+
+        if self.training:
+            x = self.kwinner_function(x, self.duty_cycle, self.k, self.boost_strength)
```

Comment:
You should get the latest `nupic.torch`. 
`kwinner_function` was refactored out. see https://github.com/numenta/nupic.torch/pull/74/

Reply:
Do you know if the changes are paramount to make it work? This is working fine on my end, and I pulled nupic.torch less than 2 weeks ago. I wouldn't like to refactor at this PR unless necessary. If it is style or performance reasons I can make the changes in the next PR.

Reply:
Did you get the latest nupic.torch from master? It does not work with the latest nupic.torch

Reply:
There is no `kwinner_function` in nupic.torch code
https://github.com/numenta/nupic.torch/search?q=kwinner_function&type=Code

Reply:
Fixed.